maths/joint_probability_distribution.py

```diff
@@ -0,0 +1,127 @@
+# https://en.wikipedia.org/wiki/Joint_probability_distribution
+# Function to calculate the joint probability distribution
+def calculate_joint_probability(
+    x_values: list, y_values: list, x_probabilities: list, y_probabilities: list
+) -> dict:
+    """
+    >>>calculate_joint_probability([1,2],[-2,5,8],[0.7,0.3],[0.3,0.5,0.2])
+    P(X=1, Y=-2) = 0.21
+    P(X=1, Y=5) = 0.35
+    P(X=1, Y=8) = 0.13999999999999999
+    P(X=2, Y=-2) = 0.09
+    P(X=2, Y=5) = 0.15
+    P(X=2, Y=8) = 0.06
+    P(X=1, Y=-2) = 0.21
+    P(X=1, Y=5) = 0.35
+    P(X=1, Y=8) = 0.13999999999999999
+    P(X=2, Y=-2) = 0.09
+    P(X=2, Y=5) = 0.15
+    P(X=2, Y=8) = 0.06
+
+    """
+    joint_distribution = {}
+
+    # Calculate the joint probability for all combinations of (X, Y)
+    for x, x_prob in zip(x_values, x_probabilities):
+        for y, y_prob in zip(y_values, y_probabilities):
+            joint_prob = x_prob * y_prob
+            joint_distribution[(x, y)] = joint_prob
+    print(joint_distribution)
+    return joint_distribution
+
+
+# Function to calculate the expectation (mean)
+def expectation(values: list, probabilities: list) -> float:
+    """
+    >>>expectation([1,2],[0.7,0.3])
+    1.2999999999999999999999
+
+    """
+    return sum(x * p for x, p in zip(values, probabilities))
+
+
+# Function to calculate the variance
+def variance(values: list, probabilities: list) -> float:
+    """
+    >>>variance([1,2],[0.7,0.3])
+    0.21000
+    """
+    mean = expectation(values, probabilities)
+    return sum((x - mean) ** 2 * p for x, p in zip(values, probabilities))
+
+
+# Function to calculate the covariance
+def covariance(
+    x_values: list, y_values: list, x_probabilities: list, y_probabilities: list
+) -> float:
+    mean_x = expectation(x_values, x_probabilities)
+    mean_y = expectation(y_values, y_probabilities)
+    return sum(
+        (x - mean_x) * (y - mean_y) * px * py
+        for x, px in zip(x_values, x_probabilities)
+        for y, py in zip(y_values, y_probabilities)
+    )
+
+
+# Function to calculate the standard deviation
+def standard_deviation(variance: list) -> float:
+    return variance**0.5
+
+
+# Input values for X and Y
```

Comment:
Please indent all of the remaining lines under `if __name__ == "__main__"` so that they will not run under pytest.